Zend/zend_alloc.c

```diff
@@ -263,6 +265,23 @@ typedef struct  _zend_mm_huge_list zend_mm_huge_list;
 
 static bool zend_mm_use_huge_pages = false;
 
+#define ZEND_MM_DEFAULT_PAGE_SIZE 4096
+
+static size_t zend_mm_get_page_size(void)
```

Comment:
I think that you can use `REAL_PAGE_SIZE` directly here

Reply:
Sorry I meant that we can use `REAL_PAGE_SIZE` directly instead of `zend_mm_get_page_size()`. `zend_alloc` relies on `REAL_PAGE_SIZE` having the right value.

Reply:
> Sorry I meant that we can use REAL_PAGE_SIZE directly instead of zend_mm_get_page_size(). zend_alloc relies on REAL_PAGE_SIZE having the right value.

Is it though? It seems to be that `REAL_PAGE_SIZE` is hardcoded to 4k, no?

```console
$ git grep 'define ZEND_MM_PAGE_SIZE'
Zend/zend_alloc_sizes.h:23:#define ZEND_MM_PAGE_SIZE  (4 * 1024)                      /* 4 KB  */
$
```

`zend_get_page_size` is used in other places as well:

```console
$ git grep zend_get_page_size
Zend/zend.c:1240:ZEND_API size_t zend_get_page_size(void)
Zend/zend.h:293:ZEND_API size_t zend_get_page_size(void);
Zend/zend_alloc.c:273:          page_size = zend_get_page_size();
Zend/zend_call_stack.c:423:     page_size = zend_get_page_size();
Zend/zend_fibers.c:193:         page_size = zend_get_page_size();
ext/opcache/ZendAccelerator.c:3198:                     page_size = zend_get_page_size();
$
```

Moreover, when I'm using `ZEND_MM_PAGE_SIZE`, things beak on a system with page size different then 4k :)

Reply:
Are you possibly mixing up `ZEND_MM_PAGE_SIZE` and `REAL_PAGE_SIZE`?

`REAL_PAGE_SIZE` expands to `_real_page_size` by default (on systems with `_SC_PAGESIZE` or `_SC_PAGE_SIZE`) and is initialized to the result of `sysconf(_SC_PAGESIZE)`

Reply:
```console
$ git grep _real_page_size
Zend/zend_alloc.c:104:#    define REAL_PAGE_SIZE _real_page_size
Zend/zend_alloc.c:105:static size_t _real_page_size = ZEND_MM_PAGE_SIZE;
$ 
```

I used `REAL_PAGE_SIZE`, and got some mprotect issues :/

Reply:
Maybe I can check. What environment were you building on?

Reply:
An Apple M2 running Asahi Linux

Reply:
Ok it will be difficult for me to check on a similar environment. Do you know if `start_memory_manager()` executes `REAL_PAGE_SIZE = sysconf(_SC_PAGESIZE)`? Maybe it needs to be adjusted.

Reply:
Found an issue: `start_memory_manager()` calls `zend_mm_init()` (via `alloc_globals_ctor()`) _before_ setting `REAL_PAGE_SIZE` to the right value. Moving this part to the top of the function should fix it:
```
#ifndef _WIN32
#  if defined(_SC_PAGESIZE)
        REAL_PAGE_SIZE = sysconf(_SC_PAGESIZE);
#  elif defined(_SC_PAGE_SIZE)
        REAL_PAGE_SIZE = sysconf(_SC_PAGE_SIZE);
#  endif
#endif
```

Reply:
I opened #14650 to fix this :)